clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java

```diff
@@ -0,0 +1,209 @@
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.errors.InvalidTopicException;
+import org.apache.kafka.common.errors.RetriableException;
+import org.apache.kafka.common.errors.TopicAuthorizationException;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.MetadataRequest;
+import org.apache.kafka.common.requests.MetadataResponse;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * <p>
+ * Manages the state of the topic metadata requests.  The manager returns the
+ * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.PollResult} when the request is ready to
+ * be sent. This manager specifically handles the following user API calls:
+ * <ul>
+ * <li>listOffsets</li>
+ * <li>partitionsFor</li>
+ * </ul>
+ * </p>
+ * <p>
+ *     The manager checks the state of the {@link CompletableTopicMetadataRequest} before sending a new one, to
+ *     prevent sending without backing off from the previous attempts.
+ *     It also checks the state of the inflight requests to prevent overwhelming the broker with duplicated requests.
+ *     The {@code inflightRequests} is memoized by the topic name.  If all topics are requested, then we use {@code
+ *     null} as the key. Once the request is completed successfully, the entry is removed.
+ * </p>
+ */
+public class TopicMetadataRequestManager implements RequestManager {
+    private final boolean allowAutoTopicCreation;
+    private final Map<String, CompletableTopicMetadataRequest> inflightRequests;
+    private final Time time;
+    private final long retryBackoffMs;
+    private final Logger log;
+
+    public TopicMetadataRequestManager(
+        final Time time,
+        final LogContext logContext,
+        final ConsumerConfig config) {
+        this.time = time;
+        this.log = logContext.logger(this.getClass());
+        this.inflightRequests = new HashMap<>();
+        this.retryBackoffMs = config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
+        this.allowAutoTopicCreation = config.getBoolean(ConsumerConfig.ALLOW_AUTO_CREATE_TOPICS_CONFIG);
+    }
+
+    @Override
+    public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
+        List<NetworkClientDelegate.UnsentRequest> requests = inflightRequests.values().stream()
+            .map(req -> req.send(currentTimeMs).orElse(null))
```

Comment:
Stylistically, could you omit the `orElse` and then the `filter` below on `isPresent`?